brain/morphology.cpp

```diff
@@ -131,14 +150,56 @@ class Morphology::Impl
             // Interpolating the cross section at point.
             const float alpha = ( length - accumLengths[index] ) /
                               ( accumLengths[index + 1] - accumLengths[index] );
-            const Vector4f sample = ( *points )[start + index + 1] * alpha +
-                                    ( *points )[start + index] * (1 - alpha );
+            const size_t start = range.first + index;
+            const Vector4f sample = ( *points )[start + 1] * alpha +
+                                    ( *points )[start] * (1 - alpha );
             result.push_back( sample );
         }
 
         return result;
     }
 
+    float getDistanceToSoma( const uint32_t sectionID ) const
+    {
+        if( _distancesToSoma.size() <= sectionID )
+            _distancesToSoma.resize( sectionID + 1 );
+
+        float& distance = _distancesToSoma[sectionID];
+        if( distance == 0)
+        {
+            // This is the soma, a first order section or the distance hasn't
+            // been computed yet. Soma and first order sections are cheap
+            // to detect and compute.
+            const int32_t parent = ( *sections )[sectionID][1];
+            if( parent == -1 || ( *types )[parent] == SECTION_SOMA )
+                return 0;
+            // For the other cases it doesn't matter to have concurrent updates
+            // because they will yield the same result (and it's probably
+            // cheaper to go ahead with the computation than to contend for a
+            // mutex).
+            distance = getSectionLength( parent ) + getDistanceToSoma( parent );
+        }
+        return distance;
+    }
+
+    floats getDistancesToSoma( const uint32_t sectionID ) const
```

Comment:
I partially agree, but I would argue that the return types are different, so you will notice you're not using the right one at compile time sooner or later. And the same happens with many of our types, e.g float vs floats.